### PR TITLE
[SymbolGraphGen] Correctly handle exported imports in swift-symbolgraph-extract

### LINKED ISSUE #59920: [SymbolGraphGen] swift-symbolgraph-extract support for emitting symbols from `@_export import`s
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15921      |
   |Radar             | rdar://problem/89687175         |
   |Original Reporter | @franklinsch      |
   |Type              | Improvement    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Swift-DocC    |
   |Labels            | Improvement        |
   |Assignee          | None      |
   |Priority          | Medium      |

   md5: 3bcd4958d4ed4ce5d66606c682cf8647

  </details>

**Issue Description:**

In https://github.com/apple/swift-docc/issues/190, we started exporting symbols that came from `@_exported import` modules when emitting symbol graph files during a Swift build. This ticket tracks adding the same support for the `swift-symbolgraph-extract`, which didn't come for free with https://github.com/apple/swift-docc/issues/190.

### PR DESCRIPTION
**Pull request by daniel-grumberg**
This sets up the necessary stuff to include symbols from exported imported modules in symbol graphs generated using the `swift-symbolgraph-extract` tool

Changes include:
- Add option to provide an allow list of reexported modules for swift-symbolgraph-extract
- Recursively collect exported imports to allow fetching all visible Decls for symbol graph generation
- Enable collection of exported imports that originate from non `SourceFile` file units for the purpose of `ModuleDecl::getDisplayDecl`

Resolves #59920 
rdar://89687175

### CODE REVIEW

**Comment by daniel-grumberg** [comment]
@swift-ci please test

**Comment by daniel-grumberg** [comment]
@swift-ci please smoke test

**Comment by daniel-grumberg** [comment]
@swift-ci please test

**Comment by daniel-grumberg** [comment]
@swift-ci please test

File: lib/AST/Module.cpp [hunk 1]
```diff
@@ -1543,40 +1509,100 @@ SourceFile::getExternalRawLocsForDecl(const Decl *D) const {
...
   }
+}
+
+void ModuleDecl::getDisplayDecls(SmallVectorImpl<Decl*> &Results, bool Recursive) const {
+  if (Recursive) {
+    ImportCollector importCollector;
+    this->getDisplayDeclsRecursivelyAndImports(Results, importCollector);
+  }
   // FIXME: Should this do extra access control filtering?
   FORWARD(getDisplayDecls, (Results));
```
Comment (daniel-grumberg) [inline lib/AST/Module.cpp:1572]: Make this an else branch

**Review by QuietMisdreavus — APPROVED** [review]
Other than the comment we spotted while walking through the code, could you add a test that verifies that the recursive export-import functionality still works? There's a similar test in `SymbolGraph/Module/ThirdOrder.swift` that you can copy to make something similar. We just need to make sure that the innermost module's symbols appear in the outer symbol graph.

**Comment by daniel-grumberg** [comment]
@swift-ci please test

**Comment by daniel-grumberg** [comment]
> Other than the comment we spotted while walking through the code, could you add a test that verifies that the recursive export-import functionality still works? There's a similar test in `SymbolGraph/Module/ThirdOrder.swift` that you can copy to make something similar. We just need to make sure that the innermost module's symbols appear in the outer symbol graph.

Test is in `test/SymbolGraph/Module/NestedExportedImport.swift`

**Review by QuietMisdreavus — APPROVED** [review]
New test looks great, thanks!

**Pull request merged**